Coscheduling: integration-test for PodGroup status

Gekko0114 opened the pull request:
#### What type of PR is this?
/kind bug
/kind feature

#### What this PR does / why we need it:
Added integration test for PodGroup Status.
Fixed some logic to ensure smooth running of the Reconcile function.

#### Which issue(s) this PR fixes:
Fixes #516 

#### Special notes for your reviewer:
Thank you for reviewing this PR.
Please let me know if I have made any mistakes or if there are any improvements that could be made.
Your comments and feedback are appreciated.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

PiotrProkop commented on test/util/utils.go:
```diff
@@ -90,6 +90,41 @@ func MakePG(name, namespace string, min int32, creationTime *time.Time, minResou
+	return pg
+}
+
+func PGStatusEquals(a, b v1alpha1.PodGroupStatus) bool {
```
you can use https://pkg.go.dev/github.com/google/go-cmp/cmp#Equal `cmp.Equal` and do not implement struct matching yourself. This package is already used in a project.

Gekko0114 replied:
Thanks! Fixed it

PiotrProkop commented on pkg/controllers/podgroup_controller.go:
```diff
@@ -102,29 +102,28 @@ func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
-	switch pgCopy.Status.Phase {
-	case "":
+
+	if pgCopy.Status.Phase == "" {
```
why change this `switch` into `if else` ?

Gekko0114 replied:
I changed the `switch-case` statements to` if-else` statements because of a reason below. 

The `switch-case` logic uses the past value of `pgCopy.Status.Phase` even if the `pgCopy.Status.Phase` changes within `switch-case` statements. As a result, we can't pass this `switch-case` statements at once, and that's why I preferred `if-else` statements. 

But this was not a bug, but a matter of preference, so I reverted to the original code. 
Thanks for your comments.

zwpaper commented on pkg/controllers/podgroup_controller.go:
it seems that it is not that necessary to edit this whole file,
1. switch works smoother than multiple if-else
2. break make there is only one patch
3. phase updating break the origin logic, is there any bugs for the origin one?

Gekko0114 commented on pkg/controllers/podgroup_controller.go:
```diff
@@ -123,8 +123,7 @@ func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 
-		if pgCopy.Status.Succeeded+pgCopy.Status.Running >= pg.Spec.MinMember &&
-			pgCopy.Status.Phase == schedv1alpha1.PodGroupScheduled {
+		if pgCopy.Status.Succeeded+pgCopy.Status.Running >= pg.Spec.MinMember {
```
I've changed this part.

In the previous code, we were only able to enter the `Running` statement when `PodGroup.Status.Phase` was set to `PodGroupScheduled`. However, in the code, only the `PostBind` logic changes `PodGroup.Status.Phase` to `PodGroupScheduled`. Due to the lack of `PostBind` logic in this test code, `PodGroup.Status.phase` cannot transition to `PodGroupRunning`.

In real cases, if the `PostBind` logic doesn't work properly, the `PodGroup` may not stop as expected. This could potentially lead to bugs.

@zwpaper, @PiotrProkop
I will appreciate your comments!

zwpaper replied:
sounds reasonable

zwpaper commented:
I am a little bit busy past few days, will make sure the testing logic works locally in the following days

Gekko0114 commented:
Thank you so much. 
Do you mean you will work on fixing the Reconcile logic or the PostBind logic? I will appreciate it.

Gekko0114 commented:
Or if you explain your idea, perhaps I can work on it instead.

zwpaper commented:
/ok-to-test

zwpaper commented:
hi @Gekko0114, can you should some test results you ran? I tried locally, and no matter I change the `want` status, it will always success.

like here https://github.com/kubernetes-sigs/scheduler-plugins/pull/549/files#diff-0a8924639ca7d77e937eeb46d44bc9a5168a94d6ffee78e0647b9e3ad4cf71b4R264 and https://github.com/kubernetes-sigs/scheduler-plugins/pull/549/files#diff-0a8924639ca7d77e937eeb46d44bc9a5168a94d6ffee78e0647b9e3ad4cf71b4R160

Gekko0114 commented:
Hi @zwpaper,
Thank you so much for your confirmation. I will double-check it.

Gekko0114 commented:
Hi @zwpaper,

I checked it again but it seems to be working fine.
When I ran this code again, all test cases passed successfully. When I changed `want` status to something else, the test failed.

I have included the logs for same test cases, and also the logs when I changed the `want` status to something else.

I tried this test code on macOS and latest repository version. 
Could you please cherry-pick my PR branch and try running the test again? If it still doesn't work, could you provide more details about your logs?

(BTW, I've updated the test code to use `gocmp.Diff`, but this is not related to this issue.)


Running successfully for same test cases.

> === RUN   TestPodGroupController/The_status_of_no_podGroup_label_pod_changes_from_pending_to_running
> I0326 22:52:17.182295   15173 podgroup_controller_test.go:281] "Creating pod " podName="t1-p1-1"
> I0326 22:52:17.184989   15173 podgroup_controller_test.go:281] "Creating pod " podName="t1-p1-2"
> I0326 22:52:17.187493   15173 podgroup_controller_test.go:281] "Creating pod " podName="t1-p1-3"
>     podgroup_controller_test.go:357: Case The status of no podGroup label pod changes from pending to running finished
> --- PASS: TestPodGroupController (6.66s)
>     --- PASS: TestPodGroupController/Statuses_of_all_pods_change_from_pending_to_running (2.92s)
>     --- PASS: TestPodGroupController/Statuses_of_all_pods_change_from_running_to_succeeded (0.89s)
>     --- PASS: TestPodGroupController/The_status_of_one_pod_changes_from_running_to_succeeded (0.85s)
>     --- PASS: TestPodGroupController/The_status_of_the_pod_changes_from_running_to_failed (0.88s)
>     --- PASS: TestPodGroupController/The_status_of_no_podGroup_label_pod_changes_from_pending_to_running (0.87s)
> PASS
> E0326 22:52:18.052442   15173 scheduling_queue.go:1072] "Error while retrieving next pod from scheduling queue" err="scheduling queue is closed"
> ok  	sigs.k8s.io/scheduler-plugins/test/integration	31.796s


Failed when changing `want` status. 
(Changing line266 to this `util.UpdatePGStatus(util.MakePG("pg1-1", ns, 3, nil, nil), "Pending", "", 0, 0, 0, 1)`)


> === RUN   TestPodGroupController/The_status_of_no_podGroup_label_pod_changes_from_pending_to_running
I0326 22:54:35.610528   15474 podgroup_controller_test.go:281] "Creating pod " podName="t1-p1-1"
I0326 22:54:35.613804   15474 podgroup_controller_test.go:281] "Creating pod " podName="t1-p1-2"
I0326 22:54:35.616372   15474 podgroup_controller_test.go:281] "Creating pod " podName="t1-p1-3"
    podgroup_controller_test.go:349:   v1alpha1.PodGroupStatus{
          	... // 3 identical fields
          	Running:   0,
          	Succeeded: 0,
        - 	Failed:    0,
        + 	Failed:    1,
          	... // 1 ignored field
          }                
    podgroup_controller_test.go:355: The status of no podGroup label pod changes from pending to running Waiting next PodGroup error: timed out waiting for the condition
--- FAIL: TestPodGroupController (16.53s)
    --- PASS: TestPodGroupController/Statuses_of_all_pods_change_from_pending_to_running (2.89s)
    --- PASS: TestPodGroupController/Statuses_of_all_pods_change_from_running_to_succeeded (0.89s)
    --- PASS: TestPodGroupController/The_status_of_one_pod_changes_from_running_to_succeeded (0.86s)
    --- PASS: TestPodGroupController/The_status_of_the_pod_changes_from_running_to_failed (0.86s)
    --- FAIL: TestPodGroupController/The_status_of_no_podGroup_label_pod_changes_from_pending_to_running (10.72s)
FAIL
E0326 22:54:46.331361   15474 scheduling_queue.go:1072] "Error while retrieving next pod from scheduling queue" err="scheduling queue is closed"
FAIL	sigs.k8s.io/scheduler-plugins/test/integration	41.908s
FAIL

zwpaper commented:
Oh my bad, I mistakenly change the args, it work after a second look.

/lgtm

zwpaper commented:
@PiotrProkop I have no more questions, is this one ok for you?

PiotrProkop commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+				st.MakePod().Namespace(ns).Name("t1-p1-3").Req(map[v1.ResourceName]string{v1.ResourceMemory: "50"}).Priority(
+					midPriority).Node(nodeName).Obj(),
+			},
+			status: []*v1alpha1.PodGroup{
```
why this field is called status when it holds the list of PodGroups? Can we get rid of this and use only `podGroups` field?

Gekko0114 replied:
I apologize for the confusion caused by my naming.

The `status` field was actually used to check the intermediate status of `podGroup`s. I have now renamed the `status` and `want` fields to `intermediatePGState` and `expectedPGState`. The `intermediatePGState` field is necessary for us to check the changes in `podGroup`'s Status.
The purpose of this test is to ensure that the `podGroup` behave as expected as the `pods` change, so we need to test the intermediate state as well.


FYI, this is the sequence of steps in the integration test:
1. Create `PodGroup` (`podGroups`)
2. Create `Pod` (`existingPods`)
3. Check intermediate status of `PodGroup` (`intermediatePGStatus`)
4. Update `Pod` (`incomingPods`)
5. Check the final status of `PodGroup` (`expectedPGStatus`)

I referred to [elasticquota_controller_test.go](https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/test/integration/elasticquota_controller_test.go) while working on this.

PiotrProkop replied:
Thanks for the explanation and rewording!

PiotrProkop commented on test/util/utils.go:
```diff
@@ -90,6 +90,19 @@ func MakePG(name, namespace string, min int32, creationTime *time.Time, minResou
 	return pg
 }
 
+func UpdatePGStatus(pg *v1alpha1.PodGroup, phase v1alpha1.PodGroupPhase, occupiedBy string, scheduled int32, running int32, succeeded int32,
```
this formatting is bugging me you should either keep all params in one line or split them and keep only one per line. You can also think about just passing status to this function instead of all possible params, like this:
```go
func UpdatePGStatus(pg *v1alpha1.PodGroup, status v1alpha1.PodGroupStatus) *v1alpha1.PodGroup{
    pg.Status = status
    return pg
}
```

PiotrProkop commented:
> @PiotrProkop,
> 
> Thank you for your feedback, I appreciate it. I've fixed it. Could you take a look again? Thanks,

Good work!
/lgtm

Gekko0114 commented:
Hi @zwpaper,
I've updated PR. Could you approve it?

denkensk commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+					t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+				}
+				if pod.Status.Phase == v1.PodRunning {
+					_, err = cs.CoreV1().Pods(pod.Namespace).UpdateStatus(testCtx.Ctx, pod, metav1.UpdateOptions{})
+					if err != nil {
+						t.Fatalf("Failed to update Pod status %q: %v", pod.Name, err)
+					}
```
```suggestion
					if _, err := cs.CoreV1().Pods(pod.Namespace).UpdateStatus(testCtx.Ctx, pod, metav1.UpdateOptions{}); err != nil {
						t.Fatalf("Failed to update Pod status %q: %v", pod.Name, err)
					}
```

denkensk commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+			// create Pods
+			for _, pod := range tt.existingPods {
+				klog.InfoS("Creating pod ", "podName", pod.Name)
+				_, err := cs.CoreV1().Pods(pod.Namespace).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
+				if err != nil {
+					t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
+				}
```
```suggestion
				if _, err := cs.CoreV1().Pods(pod.Namespace).Create(testCtx.Ctx, pod, metav1.CreateOptions{}); err != nil {
					t.Fatalf("Failed to create Pod %q: %v", pod.Name, err)
				}
```

denkensk commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+				}
+				return true, nil
+			}); err != nil {
+				t.Fatalf("%v Waiting existPods create error: %v", tt.name, err.Error())
```
```suggestion
				t.Fatalf("Waiting existPods %v create error: %v", tt.name, err.Error())
```

Gekko0114 replied:
Thanks for your comment!

tt.name is not pod name, but testname, so I think current comment is fine.

denkensk commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+					t.Fatalf("Failed to update Pod status %q: %v", pod.Name, err)
+				}
+			}
+			if err := wait.Poll(time.Millisecond*200, 10*time.Second, func() (bool, error) {
```
+ comment： // wait for all incomingPods to be scheduled

Gekko0114 replied:
Fixed it

denkensk commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+				}
+				return true, nil
+			}); err != nil {
+				t.Fatalf("%v Waiting nextPods update status error: %v", tt.name, err.Error())
```
Waiting incomingPods scheduled error？

denkensk commented on test/integration/podgroup_controller_test.go:
```diff
@@ -0,0 +1,360 @@
+				}
+				return true, nil
+			}); err != nil {
+				t.Fatalf("%v Waiting next PodGroup error: %v", tt.name, err.Error())
```
Waiting PodGroup status update error？

denkensk reviewed (commented):
Thanks @Gekko0114 
Just some small comments.

Gekko0114 commented:
Hi @denkensk,

Thanks for your comments!
I fixed it, so could you have a look again?
Thanks,

denkensk commented:
/lgtm
/approve 

Thanks for your contribution. @Gekko0114 
And sorry for the late reply.

Gekko0114 commented:
@denkensk,
I apologize for reminding you, and thank you for your prompt confirmation!

Gekko0114 commented:
/test pull-scheduler-plugins-integration-test-master

Gekko0114 commented:
When I run `make integration-test "ARGS=-run TestPodGroupController`, test can be passed.
However, `make integration-test` can't be passed... I will investigate on it.

denkensk commented:
> I apologize for reminding you, and thank you for your prompt confirmation!

@Gekko0114  It's okay, it's my responsibility. Thanks for your contribution again.

denkensk commented:
> When I run `make integration-test "ARGS=-run TestPodGroupController`, test can be passed.
> However, `make integration-test` can't be passed... I will investigate on it.

/hold  for waiting

Gekko0114 commented:
It seems that this error is caused by controller-runtime's SetupSignalHandler.
I've created an [issue](https://github.com/kubernetes-sigs/controller-runtime/issues/2263).
Please let me know if you can fix this issue.

> In this https://github.com/kubernetes-sigs/scheduler-plugins/pull/549, we have created integration test using mgr.Start(ctrl.SetupSignalHandler()). However, as [another test](https://github.com/kubernetes-sigs/scheduler-plugins/blob/master/test/integration/elasticquota_controller_test.go#L60-L76) also uses mgr.Start(ctrl.SetupSignalHandler()), we are encountering a panic [error](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_scheduler-plugins/549/pull-scheduler-plugins-integration-test-master/1642893318373249024).
> 
> panic: close of closed channel
> goroutine 9857 [running]:
> sigs.k8s.io/controller-runtime/pkg/manager/signals.SetupSignalHandler()
> /home/prow/go/src/sigs.k8s.io/scheduler-plugins/vendor/sigs.k8s.io/controller-runtime/pkg/manager/signals/signal.go:31 +0x25
> sigs.k8s.io/scheduler-plugins/test/integration.TestPodGroupController.func1()
> /home/prow/go/src/sigs.k8s.io/scheduler-plugins/test/integration/podgroup_controller_test.go:78 +0x32
> created by sigs.k8s.io/scheduler-plugins/test/integration.TestPodGroupController
> /home/prow/go/src/sigs.k8s.io/scheduler-plugins/test/integration/podgroup_controller_test.go:77 +0x3dd
> FAIL sigs.k8s.io/scheduler-plugins/test/integration 356.228s
> 
> I think this is because [ctrl.SetupSignalHandler](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/manager/signals/signal.go#L30) is being executed twice in a single go test.
> Could you suggest how we can implement integration test in such cases?

Gekko0114 commented:
Hi @denkensk ,

Could you review and approve it again?

I fixed the issue by creating this param and shared this between integration-tests.
`var signalHandler = ctrl.SetupSignalHandler()`

denkensk commented:
/lgtm
/unhold
